src/pip/_internal/req/req_file.py

```diff
@@ -230,7 +231,9 @@ def handle_option_line(
             no_index = True
             index_urls = []
         if opts.index_url and not no_index:
-            index_urls = [opts.index_url]
+            if PyPI.simple_url in index_urls:
+                index_urls.remove(PyPI.simple_url)
+            index_urls.append(opts.index_url)
```

Comment:
I think this is too difficult to make sense of if you consider the surrounding code and all possible circumstances. It may be better to modify some logic in PackageFinder so `finder.index_urls` does not automatically add the default index (`PyPI.simple_url`) eagerly, but still have it as a default if no extra indexes are added.

Reply:
This is mainly because this function is called on each line of the file without any context. The upper layer is also the generator. If you modify it according to this idea, you may need to modify it at the outermost layer

I'm sorry. Actually, I'm not familiar with the overall structure of this project.